iOS/TvShows/TvShows/Common/Network/BaseApiClient.swift

```diff
@@ -0,0 +1,164 @@
+            self.urlSession.dataTask(with: request) { [weak self] data, response, error in
+                guard
+                    let httpReponse = response as? HTTPURLResponse,
+                    let self = self,
+                    let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+                else {
+                    single(.failure(ApiError.badRequest))
+                    return
+                }
+                
+                if let error = self.mapToApiError(status: statusCode) {
+                    single(.failure(error))
+                    return
+                }
+                
+                guard let result: ResultType = self.parse(data: data) else {
+                    single(.failure(ApiError.noData))
+                    return
+                }
+
+                single(.success(result))
+            }
+            .resume()
+            
+            return Disposables.create()
+        }
+    }
+
+    private func execute<ParamsType: Encodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> Completable {
+        .create { [weak self] observer  in
+            guard let self = self else { return Disposables.create() }
+
+            let request = self.buildRequest(path: path, method: method, parameters: parameters)
+            self.urlSession.dataTask(with: request) { [weak self] _, response, error in
+                guard
+                    let httpReponse = response as? HTTPURLResponse,
+                    let self = self,
+                    let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+                else {
+                    observer(.error(ApiError.badRequest))
+                    return
+                }
+
+                if let error = self.mapToApiError(status: statusCode) {
+                    observer(.error(error))
+                    return
+                }
+
+                observer(.completed)
+
+            }
+            .resume()
+
+            return Disposables.create()
+        }
+    }
+
+    private func buildRequest<ParamsType: Encodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> URLRequest {
+        var components = URLComponents(string: "\(path)")!
+
+        if
+            method == .get || method == .delete,
+            let parameters = parameters as? [String: String]
+        {
+            components.queryItems = parameters.map(URLQueryItem.init)
+        }
+
+        var urlRequest = URLRequest(url: components.url!)
+        urlRequest.httpMethod = method.rawValue
+        urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
+
+        if
+            let parameters = parameters,
+            method == .post || method == .patch
+        {
+            urlRequest.httpBody = try? JSONSerialization.data(withJSONObject: parameters, options: .prettyPrinted)
```

Comment:
Shouldn't this also be a throwing function since parameter encoding failed ?

Reply:
This is a bit more complicated since the throwing will occur only if the conditions are met; if the method is post or patch and the parameters are not nil. In all other cases this method will require to be put in the do-catch block even thought it will never throw. I suggest leaving this for next PRs where I will refactor this method to separately handle building requests for post and patch with parameters and building requests without any of the mentioned. Is that OK with you, @kjaklinovic ?